include/cuco/detail/static_map/static_map_ref.inl

```diff
@@ -549,6 +547,232 @@ class operator_impl<
+
+      auto const group_contains_equal = group.ballot(state == detail::equal_result::EQUAL);
+      if (group_contains_equal) {
+        auto const src_lane = __ffs(group_contains_equal) - 1;
+        if (group.thread_rank() == src_lane) {
+          op(cuda::atomic_ref<T, Scope>{slot_ptr->second}, val.second);
+        }
+        return;
+      }
+
+      auto const group_contains_available = group.ballot(state == detail::equal_result::AVAILABLE);
+      if (group_contains_available) {
+        auto const src_lane = __ffs(group_contains_available) - 1;
+        auto const status   = [&, target_idx = intra_window_index]() {
+          if (group.thread_rank() != src_lane) { return insert_result::CONTINUE; }
+          return ref_.impl_.attempt_insert_stable(slot_ptr, window_slots[target_idx], val);
+        }();
+
+        switch (group.shfl(status, src_lane)) {
+          case insert_result::SUCCESS: return;
+          case insert_result::DUPLICATE: {
+            if (group.thread_rank() == src_lane) {
+              ref_.impl_.wait_for_payload(slot_ptr->second, empty_value);
+              op(cuda::atomic_ref<T, Scope>{slot_ptr->second}, val.second);
+            }
+            return;
+          }
+          default: continue;
+        }
+      } else {
+        ++probing_iter;
+      }
+    }
+  }
+
+  template <typename Value>
+  __device__ void insert_or_apply(cooperative_groups::thread_block_tile<cg_size> const& group,
+                                  Value const& value,
+                                  cuco::op::reduce::sum_tag)
+  {
+    auto& ref_ = static_cast<ref_type&>(*this);
+    ref_.insert_or_apply(group, value, [](cuda::atomic_ref<T, Scope> slot_ref, T const& payload) {
+      slot_ref.fetch_add(payload, cuda::memory_order_relaxed);
+    });
+  }
+
+ private:
+  /**
+   * @brief Attempts to insert an element into a slot or update the matching payload by applying the
+   * binary operation on the payload and new value.
+   *
+   * @tparam Value Input type which is implicitly convertible to 'value_type'
+   * @tparam Op Callable type which is used as apply operation and called be
+   *   called with arguments as Op(cuda::atomic_ref<T, Scope>, T)
+
+   * @param slot value_type pointer to the slot to insert
+   * @param value The element to insert
+   * @param op The callable object to perform binary operation between existing value at the slot
+   *  and element to insert.
+   *
+   * @return Returns `true` if the given `value` is inserted or `value` has a match in the map.
+   */
+  template <typename Value, typename Op>
+  __device__ constexpr bool attempt_insert_or_apply(value_type* slot,
+                                                    Value const& value,
+                                                    Op op) noexcept
+  {
+    ref_type& ref_          = static_cast<ref_type&>(*this);
+    auto const expected_key = ref_.impl_.empty_slot_sentinel().first;
+
+    auto old_key =
```

Comment:
Ah OK, so we are not using any of the CAS algorithms but simply CASing keys all the time. We had long discussions here thus I could miss something obvious: what are the concerns of using packed CAS in this case?

Reply:
Oops! This is a left over. We dont use `attempt_insert_or_apply` anymore, as it can only work when `sentienel == identity_element`. Because, in this situation we only need to perform **key CAS** as payload atomic operation can work in any order, as long as sentinel key is updated once. 

Currently we can only do `attempt_insert_stable` which does **stable insert** and does not rely on `identity_element`.

Reply:
could we remove `attempt_insert_or_apply` if i's no longer used?